### PR TITLE
test(function): add function e2e test verifying configs when with different plugins

### PR DESCRIPTION
**Pull request by dilin-MS**
[[e2e][function] verify configuration successfully changed when with different plugins](https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_workitems/edit/12905497)

The PR mainly adds 3 arm e2e test for function plugins.
* packages/cli/tests/e2e/function/TabAddFunctionBot.tests.ts
* packages/cli/tests/e2e/function/TabAddFunctionKeyvault.tests.ts
* packages/cli/tests/e2e/function/TabAddFunctionSql.tests.ts

Other changes contains:
* rename of method `getWebappConfigs` -> `getWebappSettings`
* Fix usage of `setSimpleAuthSkuNameToB1Bicep()` method to use `await`
* Change `FunctionValidator` from static class to non-static class so properties can be easier to be transferred in different methods
* Remove some of the if-clause for `isMultiEnvEnabled()` since we do not need them now.

### CODE REVIEW

File: packages/cli/tests/e2e/function/Arm_TabAddFunctionBot.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+  getActivePluginsFromProjectSetting,
+  getProvisionParameterValueByKey,
+  setBotSkuNameToB1Bicep,
+} from "../commonUtils";
+import { environmentManager } from "@microsoft/teamsfx-core";
+import { CliHelper } from "../../commonlib/cliHelper";
+import { Capability, Resource, provisionParametersKey } from "../../commonlib/constants";
+import { FunctionValidator } from "../../commonlib";
+
+describe("Configuration successfully changed when with different plugins", function () {
```
Comment (blackchoey) [inline packages/cli/tests/e2e/function/Arm_TabAddFunctionBot.ts:28]: can we put these test cases into 1 ts file?

File: packages/cli/tests/commonlib/functionValidator.ts [hunk 2]
```diff
@@ -152,90 +69,90 @@ export class FunctionValidator {
...
+    // Validating app settings
+    console.log("validating app settings.");
+    const webappSettingsResponse = await getWebappSettings(
+      this.subscriptionId,
+      this.rg,
+      this.functionAppName,
+      token as string
+    );
+    chai.assert.exists(webappSettingsResponse);
+    console.log("[dilin-debug] webappSettingsResponse: " + JSON.stringify(webappSettingsResponse));
```
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/commonlib/functionValidator.ts:110]: The e2e test is still under debugging. Will remove them when the E2E test passed. Thx.

File: packages/cli/tests/commonlib/functionValidator.ts [hunk 3]
```diff
@@ -28,117 +29,33 @@ enum BaseConfig {
...
-    console.log("Successfully init validator for Function.");
-    return functionObject;
+  private subscriptionId: string;
+  private rg: string;
+  private activeResourcePlugins: string[];
+  private ctx: any;
+  private functionAppName: string;
+  private resourceBaseName: string;
+
+  constructor(ctx: any, activeResourcePlugins: string[], resourceBaseName: string) {
```
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/commonlib/functionValidator.ts]: ctx can only be got *asynchronously by projectPath and env, so not suitable to move to function validator constructor.

File: packages/cli/tests/e2e/bot/Arm_Bot_CanDeployToCustomizedRg.tests.ts [hunk 4]
```diff
@@ -33,14 +33,10 @@ describe("Deploy to customized resource group", function () {
 
   const testFolder = getTestFolder();
   const subscription = getSubscriptionId();
-  let appName: string, projectPath: string;
+  const appName = getUniqueAppName();
```
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/e2e/bot/Arm_Bot_CanDeployToCustomizedRg.tests.ts:36]: Due to the e2e test time limit, each e2e test file will contains only one `describle` and one `it` test function.

File: packages/cli/tests/e2e/function/Arm_TabAddFunctionBot.ts (hunk 1 above)
Comment (dilin-MS) [reply packages/cli/tests/e2e/function/Arm_TabAddFunctionBot.ts:24]: Same as above.

File: packages/cli/tests/e2e/bot/Arm_Bot_CanDeployToCustomizedRg.tests.ts [hunk 5]
```diff
@@ -33,14 +33,10 @@ describe("Deploy to customized resource group", function () {
 
   const testFolder = getTestFolder();
   const subscription = getSubscriptionId();
```
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/e2e/bot/Arm_Bot_CanDeployToCustomizedRg.tests.ts:35]: Renamed.

File: packages/cli/tests/commonlib/functionValidator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/commonlib/functionValidator.ts]: fixed.

**Review by Alive-Fish — APPROVED** [review]

**Review by XiaofuHuang — APPROVED** [review]

File: packages/cli/tests/commonlib/cliHelper.ts [hunk 6]
```diff
@@ -113,4 +113,29 @@ export class CliHelper {
...
+      console.error(`[Failed] ${message}. Error message: ${result.stderr}`);
+    } else {
+      const arr = (result.stdout as string).split(":");
+      if (!arr || arr.length <= 1) {
+        console.error(
+          `[Failed] ${message}. Failed to get value from cli result. result: ${result.stdout}`
+        );
+      } else {
+        value = arr[1].trim() as string;
+        console.log(`[Successfully] ${message} Value: ${value}`);
```
Comment (blackchoey) [inline packages/cli/tests/commonlib/cliHelper.ts:136]: though it's for test purpose, not sure whether we should log the values, which might be secrets

File: packages/cli/tests/e2e/function/TabAddFunctionSql.tests.ts [hunk 7]
```diff
@@ -41,30 +37,26 @@ describe("Configuration successfully changed when with different plugins", funct
...
     // Provision
-    setSimpleAuthSkuNameToB1Bicep(projectPath, environmentManager.getDefaultEnvName());
-    setBotSkuNameToB1Bicep(projectPath, environmentManager.getDefaultEnvName());
+    await setSimpleAuthSkuNameToB1Bicep(projectPath, environmentManager.getDefaultEnvName());
+    await setBotSkuNameToB1Bicep(projectPath, environmentManager.getDefaultEnvName());
     await CliHelper.setSubscription(subscription, projectPath);
-    await CliHelper.provisionProject(projectPath);
+    await CliHelper.provisionProject(
+      projectPath,
+      " --sql-admin-name Abc123321 --sql-password Cab232332"
```
Comment (blackchoey) [inline packages/cli/tests/e2e/function/TabAddFunctionSql.tests.ts:45]: better not hard code password, even in test cases.

**Review by KennethBWSong — APPROVED** [review]

**Review by xzf0587 — APPROVED** [review]

File: packages/cli/tests/e2e/function/TabAddFunctionSql.tests.ts (hunk 7 above)
Comment (dilin-MS) [reply packages/cli/tests/e2e/function/TabAddFunctionSql.tests.ts:45]: Will fix in next PR to avoid dismissal of code owner's review approval. Thx

**Review by YitongFeng-git — APPROVED** [review]

File: packages/cli/tests/commonlib/functionValidator.ts [hunk 8]
```diff
@@ -285,4 +217,37 @@ export class FunctionValidator {
...
+    return expectedM365ApplicationIdUri;
+  }
+
+  private async getM365ClientSecret(
+    activeResourcePlugins: string[],
+    resourceBaseName: string
+  ): Promise<string> {
+    let m365ClientSecret: string;
+    if (activeResourcePlugins.includes(PluginId.KeyVault)) {
+      m365ClientSecret = `@Microsoft.KeyVault(VaultName=${resourceBaseName};SecretName=m365ClientSecret)`;
```
Comment (hund030) [inline packages/cli/tests/commonlib/functionValidator.ts:243]: Why is this SecretName hardcode? Can we get it from keyvault's config?

**Review by hund030 — APPROVED** [review]

File: packages/cli/tests/commonlib/functionValidator.ts (hunk 8 above)
Comment (dilin-MS) [reply packages/cli/tests/commonlib/functionValidator.ts:243]: Will fix in next PR:
Pass `keyVaultName` `m365ClientSecretName` from parameter.json file.

File: packages/cli/tests/e2e/function/TabAddFunctionBot.tests.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+  const projectPath = path.resolve(testFolder, appName);
+
+  after(async () => {
+    await cleanUp(appName, projectPath, true, false, false, true);
+  });
+
+  it(`tab + function + bot`, async function () {
+    await CliHelper.createProjectWithCapability(appName, testFolder, Capability.Tab);
+    await CliHelper.addCapabilityToProject(projectPath, Capability.Bot);
+    await CliHelper.addResourceToProject(projectPath, Resource.AzureFunction);
```
[the remark this answers is not in the document]
Comment (dilin-MS) [reply packages/cli/tests/e2e/function/TabAddFunctionBot.tests.ts:37]: Currently no priority for this. Will add test case if related bug reported.

**Review by xiaolang124 — APPROVED** [review]

**Pull request merged**